File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/AppointmentsResourceTest.java
package uk.gov.justice.hmpps.prison.api.resource.impl;

import org.junit.jupiter.api.Test;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.mock.mockito.SpyBean;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import uk.gov.justice.hmpps.prison.aop.ProxyUserAspect;
import uk.gov.justice.hmpps.prison.api.model.ScheduledEvent;
import uk.gov.justice.hmpps.prison.api.model.bulkappointments.AppointmentDefaults;
import uk.gov.justice.hmpps.prison.api.model.bulkappointments.AppointmentDetails;
import uk.gov.justice.hmpps.prison.api.model.bulkappointments.AppointmentsToCreate;
import uk.gov.justice.hmpps.prison.repository.BookingRepository;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

public class AppointmentsResourceTest extends ResourceTest {
    @SpyBean
    private ProxyUserAspect proxyUserAspect;

    @MockBean
    private BookingRepository bookingRepository;

    @Test
    public void createAnAppointment() {

        when(bookingRepository.checkBookingExists(anyLong())).thenReturn(true);
        when(bookingRepository.findBookingsIdsInAgency(any(), anyString())).thenReturn(List.of(1L, 2L));

        final var response = makeCreateAppointmentsRequest();

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);

        verify(bookingRepository).createMultipleAppointments(any(), any(), anyString());
    }

    @Test
    public void triggerProxyUserAspect() throws Throwable {
        makeCreateAppointmentsRequest();

        verify(proxyUserAspect).controllerCall(any());
    }

    @Test
    public void deleteAnAppointment() {
        final var scheduledEvent = ScheduledEvent
            .builder()
            .eventId(1L)
            .eventType("APP")
            .eventSubType("VLB")
            .startTime(LocalDateTime.of(2020, 1, 1, 1, 1))
            .endTime(LocalDateTime.of(2020, 1, 1, 1, 31))
            .eventLocationId(2L)
            .build();

        when(bookingRepository.getBookingAppointmentByEventId(anyLong())).thenReturn(Optional.of(scheduledEvent));


        final var response = testRestTemplate.exchange(
            "/api/appointments/1",
            HttpMethod.DELETE,
            createHttpEntity(validToken(List.of("ROLE_GLOBAL_APPOINTMENT")), null),
            Void.class
        );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NO_CONTENT);

        verify(bookingRepository).getBookingAppointmentByEventId(1L);
    }

    @Test
    public void deleteAnAppointment_notFound() {
        when(bookingRepository.getBookingAppointmentByEventId(1L)).thenReturn(Optional.empty());

        final var response = testRestTemplate.exchange(
            "/api/appointments/1",
            HttpMethod.DELETE,
            createHttpEntity(validToken(List.of("ROLE_GLOBAL_APPOINTMENT")), null),
            Void.class
        );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
        verify(bookingRepository).getBookingAppointmentByEventId(1L);
    }

    @Test
    public void deleteAnAppointment_notAuthorised() {
        final var response = testRestTemplate.exchange(
            "/api/appointments/1",
            HttpMethod.DELETE,
            createHttpEntity(validToken(List.of()), null),
            Void.class
        );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.FORBIDDEN);
        verifyNoInteractions(bookingRepository);
    }

    @Test
    public void getAnAppointment() {
        final var scheduledEvent = ScheduledEvent
            .builder()
            .eventId(1L)
            .eventType("APP")
            .eventSubType("VLB")
            .startTime(LocalDateTime.of(2020, 1, 1, 1, 1))
            .endTime(LocalDateTime.of(2020, 1, 1, 1, 31))
            .eventLocationId(2L)
            .build();

        when(bookingRepository.getBookingAppointmentByEventId(anyLong())).thenReturn(Optional.of(scheduledEvent));

        final var response = testRestTemplate.exchange(
            "/api/appointments/1",
            HttpMethod.GET,
            createHttpEntity(validToken(List.of("ROLE_GLOBAL_APPOINTMENT")), null),
            ScheduledEvent.class
        );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isEqualTo(scheduledEvent);
        verify(bookingRepository).getBookingAppointmentByEventId(1L);
    }

    @Test
    public void getAnAppointment_notFound() {
        when(bookingRepository.getBookingAppointmentByEventId(anyLong())).thenReturn(Optional.empty());

        final var response = testRestTemplate.exchange(
            "/api/appointments/1",
            HttpMethod.GET,
            createHttpEntity(validToken(List.of("ROLE_GLOBAL_APPOINTMENT")), null),
            String.class
        );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
        verify(bookingRepository).getBookingAppointmentByEventId(1L);
    }

    @Test
    public void getAnAppointment_notAuthorised() {
        when(bookingRepository.getBookingAppointmentByEventId(anyLong())).thenReturn(Optional.empty());

        final var response = testRestTemplate.exchange(
            "/api/appointments/1",
            HttpMethod.GET,
            createHttpEntity(validToken(List.of()), null),
            String.class
        );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.FORBIDDEN);
        verifyNoInteractions(bookingRepository);
    }

    private HttpHeaders headers(final String token, final MediaType contentType) {
        final var headers = new HttpHeaders();
        headers.setContentType(contentType);
        headers.setBearerAuth(token);
        return headers;
    }

    @Test
    public void updateAppointmentComment() {
        when(bookingRepository.updateBookingAppointmentComment(anyLong(), anyString())).thenReturn(true);

        final var response = testRestTemplate
            .exchange(
                "/api/appointments/1/comment",
                HttpMethod.PUT,
                new HttpEntity<>(
                    "Comment",
                    headers(
                        validToken(List.of("ROLE_GLOBAL_APPOINTMENT")),
                        MediaType.TEXT_PLAIN)
                ),
                Void.class
            );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NO_CONTENT);

        verify(bookingRepository).updateBookingAppointmentComment(1L, "Comment");
    }

    @Test
    public void updateAppointmentComment_emptyComment() {
        when(bookingRepository.updateBookingAppointmentComment(anyLong(), isNull())).thenReturn(true);

        final var response = testRestTemplate
            .exchange(
                "/api/appointments/1/comment",
                HttpMethod.PUT,
                new HttpEntity<>(
                    "",
                    headers(validToken(List.of("ROLE_GLOBAL_APPOINTMENT")), MediaType.TEXT_PLAIN)),
                Void.class
            );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NO_CONTENT);

        verify(bookingRepository).updateBookingAppointmentComment(1L, null);
    }

    @Test
    public void updateAppointmentComment_noComment() {
        when(bookingRepository.updateBookingAppointmentComment(anyLong(), isNull())).thenReturn(true);

        final var response = testRestTemplate
            .exchange(
                "/api/appointments/1/comment",
                HttpMethod.PUT,
                new HttpEntity<>(headers(validToken(List.of("ROLE_GLOBAL_APPOINTMENT")), MediaType.TEXT_PLAIN)),
                Void.class
            );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NO_CONTENT);

        verify(bookingRepository).updateBookingAppointmentComment(1L, null);
    }

    @Test
    public void updateAppointmentComment_notFound() {
        when(bookingRepository.updateBookingAppointmentComment(anyLong(), anyString())).thenReturn(false);

        final var response = testRestTemplate
            .exchange(
                "/api/appointments/1/comment",
                HttpMethod.PUT,
                new HttpEntity<>(
                    "Comment",
                    headers(
                        validToken(List.of("ROLE_GLOBAL_APPOINTMENT")),
                        MediaType.TEXT_PLAIN)
                ),
                Void.class
            );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);

        verify(bookingRepository).updateBookingAppointmentComment(1L, "Comment");
    }

    @Test
    public void updateAppointmentComment_unauthorised() {
        final var response = testRestTemplate
            .exchange(
                "/api/appointments/1/comment",
                HttpMethod.PUT,
                new HttpEntity<>(
                    "Comment",
                    headers(
                        validToken(List.of()),
                        MediaType.TEXT_PLAIN)
                ),
                Void.class
            );

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.FORBIDDEN);

        verifyNoInteractions(bookingRepository);
    }

    private ResponseEntity<String> makeCreateAppointmentsRequest() {
        final AppointmentsToCreate body = getCreateAppointmentBody();

        return testRestTemplate.exchange(
            "/api/appointments",
            HttpMethod.POST,
            createHttpEntity(validToken(List.of("BULK_APPOINTMENTS")), body),
            new ParameterizedTypeReference<>() {
            });
    }

    private AppointmentsToCreate getCreateAppointmentBody() {
        final var now = LocalDateTime.now().plusHours(5).truncatedTo(ChronoUnit.SECONDS);
        final var in1Hour = now.plusHours(6);
        final var bookingIds = Arrays.asList(-31L, -32L);

        final var appointments = bookingIds
            .stream()
            .map(id -> AppointmentDetails
                .builder()
                .bookingId(id)
                .comment("Comment")
                .build())
            .collect(Collectors.toList());

        final var defaults = AppointmentDefaults
            .builder()
            .locationId(-25L) // LEI-CHAP
            .appointmentType("ACTI") // Activity
            .startTime(now)
            .endTime(in1Hour)
            .build();

        return AppointmentsToCreate.builder()
            .appointmentDefaults(defaults)
            .appointments(appointments).build();
    }
}

File: src/test/java/uk/gov/justice/hmpps/prison/api/resource/impl/AppointmentsResourceTest.kt
package uk.gov.justice.hmpps.prison.api.resource.impl

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoInteractions
import org.mockito.kotlin.whenever
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.boot.test.mock.mockito.SpyBean
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
import org.springframework.http.MediaType.TEXT_PLAIN_VALUE
import org.springframework.web.reactive.function.BodyInserters
import uk.gov.justice.hmpps.prison.aop.ProxyUserAspect
import uk.gov.justice.hmpps.prison.api.model.ScheduledEvent
import uk.gov.justice.hmpps.prison.api.model.bulkappointments.AppointmentDefaults
import uk.gov.justice.hmpps.prison.api.model.bulkappointments.AppointmentDetails
import uk.gov.justice.hmpps.prison.api.model.bulkappointments.AppointmentsToCreate
import uk.gov.justice.hmpps.prison.repository.jpa.model.AgencyInternalLocation
import uk.gov.justice.hmpps.prison.repository.jpa.model.AgencyLocation
import uk.gov.justice.hmpps.prison.repository.jpa.model.EventStatus
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderBooking
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderIndividualSchedule
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderIndividualSchedule.EventClass
import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderIndividualScheduleRepository
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import java.util.Optional

class AppointmentsResourceTest : ResourceTest() {
  @SpyBean
  private lateinit var proxyUserAspect: ProxyUserAspect

  @MockBean
  private lateinit var offenderIndividualScheduleRepository: OffenderIndividualScheduleRepository

  private fun appointmentWithId(createdId1: Long): OffenderIndividualSchedule {
    val a1 = OffenderIndividualSchedule()
    a1.id = createdId1
    a1.eventType = "APP"
    a1.eventClass = EventClass.INT_MOV
    return a1
  }

  @Nested
  @DisplayName("POST /api/appointments")
  inner class CreateAppointments {
    @Test
    @Throws(Throwable::class)
    fun triggerProxyUserAspect() {
      makeCreateAppointmentsRequest(createAppointmentBody())
      verify(proxyUserAspect).controllerCall(any())
    }

    @Test
    fun createAnAppointment() {
      val firstEventId = 11
      val secondEventId = 12
      whenever(offenderIndividualScheduleRepository.save(org.mockito.kotlin.any()))
        .thenReturn(appointmentWithId(11))
        .thenReturn(appointmentWithId(12))
      val appointments = createAppointmentBody()

      val response = makeCreateAppointmentsRequest(appointments)
      response.expectStatus().isOk
      response.expectBody()
        .jsonPath("length()").isEqualTo(2)
        .jsonPath("[*].appointmentEventId").value<List<Int>> { assertThat(it).containsExactlyInAnyOrder(firstEventId, secondEventId) }
        .jsonPath("[*].bookingId").value<List<Int>> {
          assertThat(it).containsExactlyInAnyOrder(
            appointments.appointments[0].bookingId.toInt(),
            appointments.appointments[1].bookingId.toInt(),
          )
        }

      verify(offenderIndividualScheduleRepository, times(2)).save(any())
    }

    private fun makeCreateAppointmentsRequest(body: AppointmentsToCreate) =
      webTestClient.post().uri("/api/appointments")
        .headers(setAuthorisation(listOf("BULK_APPOINTMENTS")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .accept(APPLICATION_JSON)
        .body(BodyInserters.fromValue(body))
        .exchange()

    private fun createAppointmentBody(): AppointmentsToCreate {
      val now = LocalDateTime.now().plusHours(5).truncatedTo(ChronoUnit.SECONDS)
      val in1Hour = now.plusHours(6)
      val bookingIds = listOf(-31L, -32L)
      val appointments = bookingIds
        .stream()
        .map { id: Long ->
          AppointmentDetails
            .builder()
            .bookingId(id)
            .comment("Comment")
            .build()
        }
        .toList()
      val defaults = AppointmentDefaults
        .builder()
        .locationId(-25L) // LEI-CHAP
        .appointmentType("ACTI") // Activity
        .startTime(now)
        .endTime(in1Hour)
        .build()
      return AppointmentsToCreate.builder()
        .appointmentDefaults(defaults)
        .appointments(appointments).build()
    }
  }

  @Test
  fun deleteAnAppointment() {
    whenever(offenderIndividualScheduleRepository.findById(1L)).thenReturn(
      Optional.of(
        OffenderIndividualSchedule().apply {
          id = 1L
          eventType = "APP"
          eventSubType = "VLB"
          startTime = LocalDateTime.of(2020, 1, 1, 1, 1)
          endTime = LocalDateTime.of(2020, 1, 1, 1, 31)
          internalLocation = AgencyInternalLocation().apply { locationId = 2L }
          fromLocation = AgencyLocation().apply { id = "WWI" }
        },
      ),
    )

    webTestClient.delete().uri("/api/appointments/1")
      .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
      .header("Content-Type", APPLICATION_JSON_VALUE)
      .accept(APPLICATION_JSON)
      .exchange()
      .expectStatus().isNoContent
    verify(offenderIndividualScheduleRepository).findById(1L)
  }

  @Nested
  @DisplayName("DELETE /api/appointments/{appointmentId}")
  inner class DeleteAppointment {
    @Test
    fun deleteAnAppointment_notFound() {
      whenever(offenderIndividualScheduleRepository.findById(1L)).thenReturn(Optional.empty())

      webTestClient.delete().uri("/api/appointments/1")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isNotFound
      verify(offenderIndividualScheduleRepository).findById(1L)
    }

    @Test
    fun `delete an appointment not authorised`() {
      webTestClient.delete().uri("/api/appointments/1")
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isUnauthorized
      verifyNoInteractions(offenderIndividualScheduleRepository)
    }

    @Test
    fun `delete an appointment forbidden`() {
      webTestClient.delete().uri("/api/appointments/1")
        .headers(setAuthorisation(listOf("ROLE_BOB")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isForbidden
      verifyNoInteractions(offenderIndividualScheduleRepository)
    }
  }

  @Nested
  @DisplayName("GET /api/appointments/{appointmentId}")
  inner class GetAppointment {
    @Test
    fun getAnAppointment() {
      val appointment = OffenderIndividualSchedule().apply {
        id = 1L
        offenderBooking = OffenderBooking().apply { bookingId = 1L }
        eventClass = EventClass.INT_MOV
        eventType = "APP"
        eventSubType = "VLB"
        eventStatus = EventStatus("SCH", "desc")
        startTime = LocalDateTime.of(2020, 1, 1, 1, 1)
        endTime = LocalDateTime.of(2020, 1, 1, 1, 31)
        internalLocation = AgencyInternalLocation().apply { locationId = 2L }
        fromLocation = AgencyLocation().apply {
          id = "WWI"
          description = "desc"
        }
      }
      whenever(offenderIndividualScheduleRepository.findById(1L)).thenReturn(Optional.of(appointment))

      webTestClient.get().uri("/api/appointments/1")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isOk
        .expectBody(ScheduledEvent::class.java).isEqualTo(ScheduledEvent(appointment))
      verify(offenderIndividualScheduleRepository).findById(1L)
    }

    @Test
    fun getAnAppointment_notFound() {
      whenever(offenderIndividualScheduleRepository.findById(1L)).thenReturn(Optional.empty())

      webTestClient.get().uri("/api/appointments/1")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isNotFound

      verify(offenderIndividualScheduleRepository).findById(1L)
    }

    @Test
    fun `get an appointment forbidden`() {
      webTestClient.get().uri("/api/appointments/1")
        .headers(setAuthorisation(listOf("ROLE_BOB")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isForbidden
      verifyNoInteractions(offenderIndividualScheduleRepository)
    }

    @Test
    fun `get an appointment not authorised`() {
      webTestClient.get().uri("/api/appointments/1")
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isUnauthorized
      verifyNoInteractions(offenderIndividualScheduleRepository)
    }
  }

  private fun headers(token: String, contentType: MediaType): HttpHeaders {
    val headers = HttpHeaders()
    headers.contentType = contentType
    headers.setBearerAuth(token)
    return headers
  }

  @Nested
  @DisplayName("PUT /{appointmentId}/comment")
  inner class UpdateComment {
    @Test
    fun updateAppointmentComment() {
      val appointment = appointmentWithId(1)
      whenever(offenderIndividualScheduleRepository.findById(1)).thenReturn(Optional.of(appointment))

      webTestClient.put().uri("/api/appointments/1/comment")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", TEXT_PLAIN_VALUE)
        .bodyValue("Comment")
        .exchange()
        .expectStatus().isNoContent

      assertThat(appointment.comment).isEqualTo("Comment")
    }

    @Test
    fun updateAppointmentComment_emptyComment() {
      val appointment = appointmentWithId(1).apply { comment = "existing comment" }
      whenever(offenderIndividualScheduleRepository.findById(1)).thenReturn(Optional.of(appointment))

      webTestClient.put().uri("/api/appointments/1/comment")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", TEXT_PLAIN_VALUE)
        .bodyValue("")
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isNoContent

      assertThat(appointment.comment).isNull()
    }

    @Test
    fun updateAppointmentComment_noComment() {
      val appointment = appointmentWithId(1).apply { comment = "existing comment" }
      whenever(offenderIndividualScheduleRepository.findById(1)).thenReturn(Optional.of(appointment))

      webTestClient.put().uri("/api/appointments/1/comment")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", TEXT_PLAIN_VALUE)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isNoContent

      assertThat(appointment.comment).isNull()
    }

    @Test
    fun updateAppointmentComment_notFound() {
      whenever(offenderIndividualScheduleRepository.findById(1)).thenReturn(Optional.empty())

      webTestClient.put().uri("/api/appointments/1/comment")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", TEXT_PLAIN_VALUE)
        .bodyValue("Comment")
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isNotFound
    }

    @Test
    fun `update appointment comment forbidden`() {
      webTestClient.put().uri("/api/appointments/1/comment")
        .headers(setAuthorisation(listOf("ROLE_BOB")))
        .header("Content-Type", TEXT_PLAIN_VALUE)
        .bodyValue("")
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isForbidden

      verifyNoInteractions(offenderIndividualScheduleRepository)
    }

    @Test
    fun `update appointment comment not authorised`() {
      webTestClient.put().uri("/api/appointments/1/comment")
        .header("Content-Type", TEXT_PLAIN_VALUE)
        .bodyValue("")
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isUnauthorized

      verifyNoInteractions(offenderIndividualScheduleRepository)
    }
  }

  @Nested
  @DisplayName("PUT /{appointmentId}/comment/v2")
  inner class UpdateCommentV2 {
    @Test
    fun updateAppointmentComment() {
      val appointment = appointmentWithId(1)
      whenever(offenderIndividualScheduleRepository.findById(1)).thenReturn(Optional.of(appointment))

      webTestClient.put().uri("/api/appointments/1/comment/v2")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .bodyValue("""{ "comment": "Comment" } """)
        .exchange()
        .expectStatus().isNoContent

      assertThat(appointment.comment).isEqualTo("Comment")
    }

    @Test
    fun updateAppointmentComment_emptyComment() {
      val appointment = appointmentWithId(1).apply { comment = "existing comment" }
      whenever(offenderIndividualScheduleRepository.findById(1)).thenReturn(Optional.of(appointment))

      webTestClient.put().uri("/api/appointments/1/comment/v2")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .bodyValue("""{ "comment": "" } """)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isNoContent

      assertThat(appointment.comment).isNull()
    }

    @Test
    fun updateAppointmentComment_noComment() {
      val appointment = appointmentWithId(1).apply { comment = "existing comment" }
      whenever(offenderIndividualScheduleRepository.findById(1)).thenReturn(Optional.of(appointment))

      webTestClient.put().uri("/api/appointments/1/comment/v2")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isNoContent

      assertThat(appointment.comment).isNull()
    }

    @Test
    fun updateAppointmentComment_notFound() {
      whenever(offenderIndividualScheduleRepository.findById(1)).thenReturn(Optional.empty())

      webTestClient.put().uri("/api/appointments/1/comment/v2")
        .headers(setAuthorisation(listOf("ROLE_GLOBAL_APPOINTMENT")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .bodyValue("""{ "comment": "Comment" } """)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isNotFound
    }

    @Test
    fun `update appointment comment forbidden`() {
      webTestClient.put().uri("/api/appointments/1/comment/v2")
        .headers(setAuthorisation(listOf("ROLE_BOB")))
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .bodyValue("""{ "comment": "Comment" } """)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isForbidden

      verifyNoInteractions(offenderIndividualScheduleRepository)
    }

    @Test
    fun `update appointment comment not authorised`() {
      webTestClient.put().uri("/api/appointments/1/comment/v2")
        .header("Content-Type", APPLICATION_JSON_VALUE)
        .bodyValue("""{ "comment": "Comment" } """)
        .accept(APPLICATION_JSON)
        .exchange()
        .expectStatus().isUnauthorized

      verifyNoInteractions(offenderIndividualScheduleRepository)
    }
  }
}
